lib/src/embedded/isolate_dispatcher.dart

```diff
@@ -70,13 +72,15 @@ class IsolateDispatcher {
         (compilationId, messageBuffer) = parsePacket(packet);
 
         if (compilationId != 0) {
-          // TODO(nweiz): Consider using the techniques described in
-          // https://github.com/dart-lang/language/issues/124#issuecomment-988718728
-          // or https://github.com/dart-lang/language/issues/3118 for low-cost
-          // inter-isolate transfers.
-          (_activeIsolates[compilationId] ?? await _getIsolate(compilationId))
-              .add(messageBuffer);
-          return;
+          var mailbox =
+              (_mailboxes[compilationId] ?? await _getMailbox(compilationId));
+          try {
+            mailbox.put(packet);
+            return;
+          } on StateError catch (_) {
+            throw paramsError(
+                "InboundMessage with compilation ID $compilationId is out of order.");
```

Comment:
This is a limit of mailbox can only hold one message at a time, meaning any speculative message send from the compiler will cause issue.

Imagine the following sequence:
1. Host send compile request
2. Compiler send function call request
3. Host send function call response
4. Compiler send compiler response

If the mailbox is a synchronized queue, then it will be possible to write a test where:

1. Host send compile request
2. Host send function call response (buffered in the queue)
3. Compiler send function call request (response consumed from queue)
4. Compiler send compiler response

This kind of speculative messaging is not going to work due to mailbox is limited to single message.